pkg/3scale/amp/component/system.go

```diff
@@ -767,6 +756,33 @@ func (system *System) FileStorageVolume() v1.Volume {
 	}
 }
 
+func (system *System) s3FileStorageVolume() v1.Volume {
+	return v1.Volume{
+		// TODO here we could use the secret name provided as the volume name but
+		// this means if the secret name changes the value that arrives here would
+		// change too. Do we really want to allow that to happen/reconcile it?
+		Name: SystemFileStorageS3VolumeName,
+		VolumeSource: v1.VolumeSource{
+			Secret: &v1.SecretVolumeSource{
+				SecretName: system.Options.s3FileStorageOptions.S3ConfigSecret,
+			},
+		},
+	}
+}
+
+func (system *System) fileStorageVolume() v1.Volume {
+	if system.Options.pvcFileStorageOptions != nil {
+		return system.pvcFileStorageVolume()
+	}
+
+	if system.Options.s3FileStorageOptions != nil {
+		return system.s3FileStorageVolume()
+	}
+
+	// TODO should we panic? what would be best approach here?
+	panic("One of the volume types should be defined")
```

Comment:
See comment above

Reply:
I do not think this is the correct approach. 

This kind of logic belongs to `operator` or `template` packages. `component` packages should only contain functions to generate OCP objects. 

High level there should be: `func s3AppDeploymentConfig()` and `func pvcAppDeploymentConfig` and make all generalizations (to leverage code reuse) downstream until you get your implementation. In other words, the volume could be created and passed to a function that creates deploymentconfig and uses the param to populate `VolumeMounts`.

Reply:
I have concerns about the proposed approach as you are going the way the same approach as the templates take. Once you start to have multiple combinations you would have to implement each one of those and start to name them differently for each combination, even if you reuse code:

```
AppDeploymentConfigWithS3AsFileStorage
AppDeploymentConfigWithPVCAsFileStorage
AppDeploymentConfigWithS3AsFileStorageAndAnotherSpecificConfig1
AppDeploymentConfigWithS3AsFileStorageAndAnotherSpecificConfig2
AppDeploymentConfigWithPVCAsFileStorageAndAnotherSpecificConfig1
AppDeploymentConfigWithPVCAsFileStorageAndAnotherSpecificConfig2
```

repeat for the N possible combinations among them.